### ISSUE TITLE
Fabio -> HTTPS -> Service ?

### ISSUE THREAD

**Issue by BogetC** [fabiolb/fabio#503]
Hi !
I need to have 100% secure comunications.
Is it possible to forward request to backends with HTTPS ?

and also, is fabio still maintain

**Comment by magiconair** [comment]
You can use the tcp+sni proxy mode to have full end-to-end encryption. 

Yes, fabio is still maintained. However, I am currently relocating my family to another country which uses up a lot of my free time. 

—
Frank Schröder

**Comment by BogetC** [comment]
Thanks for this quick answer !

I will have a look at that. I'll keep you update.

Thanks again.

**Comment by BogetC** [comment]
Hi again ! 

I'm trying to store certificates in consul, and would like fabio to send the right certificate when a request is forwarded. 
How tell fabio to pick the right certificates in consul to forward HTTPS request ? Is there a keyword to bind fabio and consul certificates ?

**Comment by BogetC** [comment]
Ping.

Hello again.

I'm getting crazy. Fabio is great but for one last feature i need, i cannot move forward.

My services have to listen in https. Is there a way to configure fabio to send https request ? 

tcp+sni proto seems to be to much constraint. i get : **[WARN] No route for IP**
tlsskipverify doesn't work for me. i get :  **tls: client didn't provide a certificate**
and i get this last error with every config i try.

I think the documentation needs more examples.

**Comment by BogetC** [comment]
@magiconair I could need your help. 
Sorry to spam this issue but there is not enought documentation and not enought place to talk about fabio.

**Comment by leprechau** [comment]
You basically have two options for end-to-end TLS communication:

1.  `tcp+sni` listener will route to the desired backend based on the SNI header of the incoming request and do a simple TCP proxy.  In this case fabio does not terminate TLS and leaves it up to the backend to terminate the TLS connection with the proper certificate chain.

1.  `https` listener will route to the desired backend based on host, path or port just like a regular HTTP listener.  In this case fabio will terminate the TLS connection and then re-originate optionally also  https to a backend system.

**Comment by BogetC** [comment]
with https listener i get : 

for my service : **tls: client didn't provide a certificate**

for fabio **http: proxy error: remote error: tls: bad certificate**

**Comment by leprechau** [comment]
To get mostly end-to-end TLS with an `https` type listener you should add the tags as described in the documentation to your backend services:

https://github.com/fabiolb/fabio/wiki/Features#https-upstream-support

Keep in mind the backend must support terminating HTTPS and provide proper certificates.  You may skip validation if needed.

**Comment by BogetC** [comment]
done

**Comment by BogetC** [comment]
skip validation doesn't work

my service tags are : urlprefix-/A proto=https tlsskipverify=true

**Comment by BogetC** [comment]
even with that i get : **tls: client didn't provide a certificate**

**Comment by leprechau** [comment]
What does your routing table look like?  Could you post the output of `curl localhost:9998/api/routes?raw`?

**Comment by leprechau** [comment]
So, this seems like your backend isn't offering HTTPS or you are making an HTTPS request to an HTTP port.

**Comment by BogetC** [comment]
route add s2 /C https://137.74.25.188:8002 opts "tlsskipverify=true"
route add s1 /B https://137.74.25.192:8001
route add s2 /A https://137.74.25.188:8002 opts "tlsskipverify=true"
route add s1 /A https://137.74.25.192:8001

**Comment by BogetC** [comment]
My backend is right in golang, and seems to be ok but i ll have a look at that to be sure

**Comment by leprechau** [comment]
I don't see the `proto=https` option there.  Secondly, are you doing TLS on port 8002?

**Comment by BogetC** [comment]
My service tags are : **urlprefix-/A proto=https tlsskipverify=true", "urlprefix-/C proto=https tlsskipverify=true**

**Comment by BogetC** [comment]
yeah for 8002, which port should i use ? 443 is already in use

**Comment by leprechau** [comment]
That's fine, just making sure I was making the correct assumption.  The routing table looks a bit odd then.  You have `https` destinations going to both port `8001` and `8002` and I am assuming that port 8001 is plain HTTP.  So my guess is that you have `proto=https` on both services? The 8001 service and the 8002 service?

**Comment by BogetC** [comment]
yeah but they are both running on different machines

**Comment by leprechau** [comment]
So for clarification ... is this annotation correct?

service listening for HTTPS with a self-signed cert on port 8002
```
route add s2 /C https://137.74.25.188:8002 opts "tlsskipverify=true"
```
service listening for HTTPS with a valid cert on port 8001
```
route add s1 /B https://137.74.25.192:8001
```
service listening for HTTPS with a self-signed cert on port 8002
```
route add s2 /A https://137.74.25.188:8002 opts "tlsskipverify=true"
```
service listening for HTTPS with a valid cert on port 8001

```
route add s1 /A https://137.74.25.192:8001
```

**Comment by BogetC** [comment]
Server 1 (s1):  implements service A and B. Listen for HTTPS with self signed cert 

Same for Server 2 (s2): just implements C instead of B

**Comment by BogetC** [comment]
Only one contain tlsskipverify, only for testing

**Comment by leprechau** [comment]
Okay but I'm specifically concerned about what port the services listen on.  If port 8001 is HTTP your routing table is wrong.

**Comment by BogetC** [comment]
I would them both doesn't use tlsskipverify

**Comment by BogetC** [comment]
why 8001 is wrong ?

**Comment by BogetC** [comment]
i'm using **server.ListenAndServeTLS(certPath, keyPath)** for both backends

**Comment by leprechau** [comment]
It's only wrong if port 8001 is an HTTP listener.  If both port 8001 and port 8002 terminate TLS then it's fine.

Could you test a curl?

```
curl --head -v -k https://137.74.25.192:8001
```
and ...
```
curl --head -v -k https://137.74.25.188:8002
```

**Comment by BogetC** [comment]
i have to go, but i ll be back in 45min. Is it possible to join you with facebook or something to continue this discussion ? 

I could post the entire solution if we find it after.

**Comment by leprechau** [comment]
I will be here.

**Comment by BogetC** [comment]
* Rebuilt URL to: https://137.74.25.192:8001/
*   Trying 137.74.25.192...
* TCP_NODELAY set
* Connected to 137.74.25.192 (137.74.25.192) port 8001 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS alert, Server hello (2):
* error:14094412:SSL routines:SSL3_READ_BYTES:sslv3 alert bad certificate
* stopped the pause stream!
* Closing connection 0
curl: (35) error:14094412:SSL routines:SSL3_READ_BYTES:sslv3 alert bad certificate

**Comment by BogetC** [comment]
* Rebuilt URL to: https://137.74.25.192:8001/
*   Trying 137.74.25.192...
* TCP_NODELAY set
* Connected to 137.74.25.192 (137.74.25.192) port 8001 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Request CERT (13):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS alert, Server hello (2):
* error:14094412:SSL routines:SSL3_READ_BYTES:sslv3 alert bad certificate
* stopped the pause stream!
* Closing connection 0
curl: (35) error:14094412:SSL routines:SSL3_READ_BYTES:sslv3 alert bad certificate
MacBook-Air-de-Corentin-2:ssl corentinboget$ curl --head -v -k https://137.74.25.188:8002
* Rebuilt URL to: https://137.74.25.188:8002/
*   Trying 137.74.25.188...
* TCP_NODELAY set
* Connection failed
* connect to 137.74.25.188 port 8002 failed: Connection refused
* Failed to connect to 137.74.25.188 port 8002: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 137.74.25.188 port 8002: Connection refused

**Comment by BogetC** [comment]
Ok great, I ll be back as soon as i can. thanks for helping me. You are my last hope x)

**Comment by leprechau** [comment]
Okay, that's pretty informative.  The certificate you are providing isn't parsable for `137.74.25.192:8001` and it doesn't look like `137.74.25.188` is even listening on port `8002`.

**Comment by BogetC** [comment]
back online @leprechau

**Comment by BogetC** [comment]
ok i understand what you said.
Let me check on that. 

Thanks for your time ;)

**Comment by BogetC** [comment]
@leprechau . How do i tell fabio to provide cert for 137.74.25.192 ?

**Comment by BogetC** [comment]
yeah i just switch port 8002 to 8001 to show you that is two backends on different machines

**Comment by leprechau** [comment]
Understood about the ports.  To the certificate issue...

With an HTTPs listener fabio is going to terminate the TLS handshake from the client and then re-originate a request to a backend server.

Using an `https` fabio listener and a standard backend (no special `proto` option) ...
```
[client] <--- HTTPs ---> [fabio] <---- HTTP ----> [backend]
```

Using an `https` listener with an `https` backend declared with `proto=https` ...
```
[client] <--- HTTPs ---> [fabio] <---- HTTPS ----> [backend]
```

In each situation fabio is acting as a server (handling request from client) and as a client (making request to the backend).  When you define the backend as an `https` endpoint that backend needs to be able to terminate TLS requests.  Fabio will provide the certificate for the original client connection (frontend) but it can not provide the certificate for the backend as it's acting as a client to the backend.  The backend system will need to provide the certificate and handle the TLS termination.  This is not specific to fabio ... this is the basic design and flow for any reverse proxy.

In short, if you cannot make an `https` curl request to the backend fabio will also not be able to make an `https` request to the backend system.

**Comment by BogetC** [comment]
Hi @leprechau.

I re-generate my certificates. 

I succeed contact my backends with postman and the right client certificate. But when i'm trying to contact fabio with the same certificate:  

fabio tells me: **http: proxy error: remote error: tls: bad certificate**
backend tells me: **http: TLS handshake error from 137.74.25.173:38006: tls: client didn't provide a certificate**

so the same error than yesterday.

I am confuse about the issue. If you have any idea, it could help me.
Thanks for your time.

**Comment by BogetC** [comment]
If my backends have to provide certificates, why are they tell me that client didn't provide it ?

**Comment by BogetC** [comment]
When i use curl with client certificate to contact my backends it work. I don't understand how my backends could provide certificate for client.

**Comment by BogetC** [comment]
curl -v --cacert myCa.crt --key certs/myComputer.key --cert certs/myComputer.crt https://137.74.25.192:8001/B

This worked.

**Comment by BogetC** [comment]
The probleme comes with Fabio so let me know if there is something i miss. 
It seems like add proto=https to the the backends tags is not enought.
